### ISSUE TITLE
Unreliable decryption with Obok DeDRM

### ISSUE THREAD

**Issue by Zalugar** [apprenticeharper/DeDRM_tools#2329]
Calibre: Version 7.12.0
Kobo Desktop: Version 4.37.19051
DeDRM: Version 10.0.9

Issue: When decrypting batches of epubs bought from from Kobo and downloaded via the Kobo Desktop, there's always 1-2 books that struggle to decrypt. Out of 20 books (I buy batches monthly), 1-2 will fail to decrypt and be unreadable every time, even though the other majority will decrypt without issue. _I can usually fix the ones that fail to decrypt by deleting the Kobo files and redownloading them until it magically works_ once out of 100 times. The decrypter will randomly think it's a green lock or red lock between tries, but this doesn't seem to have any bearing on the whether it's going to successfully decrypt it or not. Whether it succeeds or not, it thinks it succeeds regardless (says no issues).

This has been a persistent issue ever since Kobo removed the ability to download epub links from your website account to download through Adobe Digital Editions. Here's a log of me trying to decrypt a book and then trying to open it

## Log
```
ApplicationPaletteChange event ignored
Using calibre Qt style: True
ApplicationPaletteChange event received
calibre Debug log
calibre 7.12  embedded-python: True
Windows-10-10.0.22621-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22621')
Python 3.11.5
Windows: ('10', '10.0.22621', 'SP0', 'Multiprocessor Free')
Interface language: None
EXE path: C:\Program Files\Calibre2\calibre-debug.exe
Successfully initialized third party plugins: DeDRM (10, 0, 9) && Obok DeDRM (10, 0, 9)
calibre 7.12  embedded-python: True
Windows-10-10.0.22621-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22621')
Python 3.11.5
Windows: ('10', '10.0.22621', 'SP0', 'Multiprocessor Free')
Interface language: None
EXE path: C:\Program Files\Calibre2\calibre-debug.exe
Successfully initialized third party plugins: DeDRM (10, 0, 9) && Obok DeDRM (10, 0, 9)
QPA platform: windows
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 108.37333333333332 x 107.5764705882353
[0.00] Starting up...
[0.00] Showing splash screen...
[0.16] splash screen shown
[0.16] Initializing db...
[0.16] db initialized
[0.16] Constructing main UI...
DEBUG:    0.0 obok::utilities.py - loading translations
DEBUG:    0.0 obok::dialogs.py - loading translations
DEBUG:    0.0 obok::config.py - loading translations
DEBUG:    0.0 obok::action_err.py - loading translations
[0.39] GUI main window shown
[1.23] main UI initialized...
[1.23] Hiding splash screen
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 481, in get_file_type_metadata
  File "calibre\customize\builtins.py", line 177, in get_metadata
  File "calibre\ebooks\metadata\epub.py", line 216, in get_metadata
  File "calibre\ebooks\metadata\epub.py", line 154, in read_bytes
  File "calibre\utils\zipfile.py", line 1005, in read
  File "calibre\utils\zipfile.py", line 1033, in open
  File "calibre\utils\zipfile.py", line 994, in getinfo
KeyError: "There is no item named 'OPS/advanced.opf' in the archive"
```

**Comment by ElleKayEm** [comment]
That book looks like the sort of book that calibre's viewer might choke on because there really isn't an advance.opf and some place in the book says there should be.  Some other apps will ignore this.

If it is showing a green open lock, Obok will import it with no attempt to decrypt because it thinks it is already DRM-free.  In some cases, of course, it actually is and that's what should happen.  Always have Kobo Desktop closed before importing and try smaller batches at a time.

Curious which Kobo site doesn't allow downloads from the website anymore and when did that happen?  Kobo US still does.

**Comment by Zalugar** [comment]
Closing Kobo Desktop before importing seems to have made it a more consistent red lock (it was randomizing between green and red locks per redownload). But I've been testing with both just 1 file and sometimes 2 files since I removed the successful ones after the first go, both methods which have the same result of no success.

It is Kobo US. They used to allow me to download what I bought from Rakuten Kobo (the Japanese site) but Kobo US stopped allowing this feature about half a year ago (the Japanese site never let you download them so I had to the US site to do so since they're synced); I imagine it might let me download US Kobo-bought books, but I don't care to test this theory.
